include/fi_tagged_rx.h

```diff
@@ -0,0 +1,156 @@
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _FI_TAGGED_RX_H_
+#define _FI_TAGGED_RX_H_
+
+#include <fi.h>
+#include <fi_mem.h>
+#include <rdma/fi_tagged.h>
+
+/*
+ * Tagged recv list template: provides a set of utility api's for
+ * software-level tag matching, managing recv entries and unexpected
+ * messages.
+ * 
+ * Caller must provider an entrytype of the following
+ * structure:
+ *
+ * struct <entry_type> {
+ *         <mandatory fields>
+ *         struct dlist_entry entry;
+ *         struct fi_msg_tagged tmsg;
+ *         int fs_entry;
+ *
+ *         <optional: iovec's and other field(s)>
+ *         struct iovec iov[MAX_IOV_LIMIT];
+ *         void *desc[MAX_IOV_LIMIT];
+ * };
+ *
+ * Usage: 
+ *
+ * <name>_create - Create tagged-recv list, with a maximum of
+ * `num_rx' recv entries in freestack
+ *
+ * <name>_free - Frees the tagged-recv list
+ *
+ * <name>_entry_alloc - gets an entry from freestack. Caller can enqueue
+ * it to the posted/unexpected list as needed
+ *
+ * <name>_search_post - search for a matching recv entry with the given
+ * tag/addr in posted list
+ * 
+ * <name>_serch_unexp - search for a matching recv entry with the given
+ * tag/ignore/addr in unexpected list
+ */
+#define DECLARE_TAGGED_RX(entrytype, name)				\
+	DECLARE_FREESTACK(entrytype, name ## _fs)			\
+	struct name {							\
+		struct dlist_entry post_list;				\
+		struct dlist_entry unexp_list;				\
+		struct name ##_fs *fs;					\
+	};								\
+									\
+	static inline struct name *name ## _create(ssize_t num_rx)	\
+	{								\
+		struct name *list;					\
+		list = calloc(1, sizeof(struct name));			\
+		if (list) {						\
+			list->fs = name ## _fs_create(num_rx);		\
+				dlist_init(&list->post_list);		\
+				dlist_init(&list->unexp_list);		\
+		}							\
+		return list;						\
+	}								\
+									\
+	static inline void name ## _free(struct name *list)		\
+	{								\
+		name ## _fs_free(list->fs);				\
+		free(list);					        \
+	}								\
+									\
+	static inline entrytype *name ## _entry_alloc(struct name *list)\
+	{								\
+		int i;							\
+		entrytype *entry;					\
+		if (freestack_isempty(list->fs))			\
+			return NULL;					\
+		i = freestack_pop(list->fs);				\
+		entry = &list->fs->buf[i];				\
+		entry->fs_entry = i;					\
```

Comment:
I would rename fs_entry to index.  Please initialize on creation, to avoid setting this on each alloc